crypto/ml_dsa/ml_dsa_key.c

```diff
@@ -0,0 +1,436 @@
+    }
+    return 0;
+}
+
+/*
+ * @brief Given a key containing private key values for rho, s1 & s2
+ * generate the public value t and return the compressed values t1, t0.
+ *
+ * @param key A private key containing params, rh0, s1 & s2.
+ * @param md_ctx A EVP_MD_CTX used for sampling.
+ * @param t1 The returned polynomial encoding of the 10 MSB of each coefficient
+ *        of the uncompressed public key polynomial t.
+ * @param t0 The returned polynomial encoding of the 13 LSB of each coefficient
+ *        of the uncompressed public key polynomial t.
+ * @returns 1 on success, or 0 on failure.
+ */
+static int public_from_private(const ML_DSA_KEY *key, EVP_MD_CTX *md_ctx,
+                               VECTOR *t1, VECTOR *t0)
+{
+    const ML_DSA_PARAMS *params = key->params;
+    uint32_t k = params->k, l = params->l;
+    POLY *polys;
+    MATRIX a_ntt;
+    VECTOR s1_ntt;
+    VECTOR t;
+
+    polys = OPENSSL_malloc(sizeof(*polys) * (k + l + k * l));
+    if (polys == NULL)
+        return 0;
+
+    vector_init(&t, polys, k);
+    vector_init(&s1_ntt, t.poly + k, l);
+    matrix_init(&a_ntt, s1_ntt.poly + l, k, l);
+
+    /* Using rho generate A' = A in NTT form */
+    if (!matrix_expand_A(md_ctx, key->shake128_md, key->rho, &a_ntt))
+        goto err;
+
+    /* t = NTT_inv(A' * NTT(s1)) + s2 */
+    vector_copy(&s1_ntt, &key->s1);
+    vector_ntt(&s1_ntt);
+
+    matrix_mult_vector(&a_ntt, &s1_ntt, &t);
+    vector_ntt_inverse(&t);
+    vector_add(&t, &key->s2, &t);
+
+    /* Compress t */
+    vector_power2_round(&t, t1, t0);
+
+    /* Zeroize secret */
+    vector_zero(&s1_ntt);
+err:
+    OPENSSL_free(polys);
+    return 1;
+}
+
+static int ml_dsa_key_public_from_private(ML_DSA_KEY *key)
+{
+    int ret = 0;
+    VECTOR t0;
+    EVP_MD_CTX *md_ctx = NULL;
+
+    if (!vector_alloc(&t0, key->params->k)) /* t0 is already in the private key */
+        return 0;
+    if (!ossl_ml_dsa_key_pub_alloc(key)) /* allocate space for t1 */
+        return 0;
+
+    md_ctx = EVP_MD_CTX_new();
+    if (md_ctx == NULL)
+        goto err;
+
+    ret = public_from_private(key, md_ctx, &key->t1, &t0)
+        && ossl_ml_dsa_pk_encode(key)
+        && shake_xof(md_ctx, key->shake256_md,
+                     key->pub_encoding, key->params->pk_len,
+                     key->tr, sizeof(key->tr));
+err:
+    vector_free(&t0);
+    EVP_MD_CTX_free(md_ctx);
+    return ret;
+}
+
+
+/**
+ * @brief Load a ML_DSA key from raw data.
+ *
+ * @param key An ML_DSA key to load into
+ * @param params An array of parameters containing key data.
+ * @param include_private Set to 1 to optionally include the private key data
+ *                        if it exists.
+ * @returns 1 on success, or 0 on failure.
+ */
+int ossl_ml_dsa_key_fromdata(ML_DSA_KEY *key, const OSSL_PARAM params[],
+                             int include_private)
+{
+    const OSSL_PARAM *p = NULL, *priv = NULL;
+
+    p = OSSL_PARAM_locate_const(params, OSSL_PKEY_PARAM_PUB_KEY);
```

Comment:
This makes it possible to load an inconsistent key pair, where the public and private don't match. I think the private should preëmpt the public, and always have the public part re-generated.

Reply:
Its an arbitrary decision. If they specify the public key, it should probably load it.
Worst case is the verify fails right?

Reply:
Loading an _inconsistent_ key pair is **extremely unlikely**.

I don't think we should many any special effort to detect or mitigate this case.  The primary reason being _consistency_.  Nowhere else do we generate the public from the private when both are offered.  Yes, I know about ML-KEM and it should be bought into line.

If you believe that we should generate the public from the private ignoring the offered public key, feel free to raise an OTC discussion and/or an issue or PR.  We should make the change _universally_ rather than promote glaringly inconsistent behaviour.

Reply:
I am no way riled up, but rather bemused by the tone of this response. There's no "glaringly" here, and I don't think this is an OTC-level issue, but perhaps the OTC does indeed engage at this layer of detail much to my surprise.

Anyway, inconsistent keys can arise not only during initial import in `fromdata` with *both* offered, but also as a result of much later *augmentation*, with the key already "live". This is my primary concern, because the interface does not prevent adding components to keys already in use. Such keys should not be mutated.

And yes, I also firmly believe that we should avoid inconsistent states, and so load private keys preferentially when both are provided, deriving the public if that's

Reply:
I have changed the code, it also checks that it you supply the public key that it matches what was generated. otherwise it errors. ML_KEM should also work like this.